File: mobiler/plugins/analytics/ios/AnalyticsPlugin.swift
import FirebaseAnalytics
import FirebaseCore
import FirebaseCrashlytics
import Foundation
import SharedTypes

// analytics (free, bundled, EXPERIMENTAL). Product analytics + automatic crash capture via Firebase
// Analytics + Crashlytics. Fire-and-forget request/response (no stream). Crash capture is AUTOMATIC once
// Firebase is configured. Ops (input is JSON unless noted):
//   logEvent {name, params?}   setUserId <id>   setUserProperty {name, value}   setEnabled "true"|"false"
//   log <message>              recordError {message, domain?}                    testCrash ""  (dev only)
//
// `AnalyticsPlugin.bootstrap()` runs at launch (App.swift `// mobiler:app-launch`) and configures
// Firebase IF a GoogleService-Info.plist is bundled. Without it, the app does NOT crash — every op
// short-circuits to ok:false ("Firebase not configured") so the plugin is safe to add before config.
enum AnalyticsPlugin {
    // Launch hook — configure Firebase early so Crashlytics catches startup crashes. Guarded twice:
    // skip if already configured (push-firebase-only may have), and skip if no config file is present.
    static func bootstrap() {
        guard FirebaseApp.app() == nil else { return }
        guard Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil else { return }
        FirebaseApp.configure()
    }

    static func handle(op: String, input: String) async -> PluginResponse {
        guard FirebaseApp.app() != nil else {
            return PluginResponse(ok: false, output: "Firebase not configured — add GoogleService-Info.plist")
        }
        switch op {
        case "logEvent": return logEvent(input)
        case "setUserId":
            Analytics.setUserID(input.isEmpty ? nil : input)
            return PluginResponse(ok: true, output: "")
        case "setUserProperty": return setUserProperty(input)
        case "setEnabled": return setEnabled(input)
        case "log":
            Crashlytics.crashlytics().log(input)
            return PluginResponse(ok: true, output: "")
        case "recordError": return recordError(input)
        case "testCrash":
            // Deliberate crash to verify Crashlytics wiring — the report uploads on the NEXT launch.
            fatalError("analytics testCrash")
        default:
            return PluginResponse(ok: false, output: "unknown op '\(op)'")
        }
    }

    private static func logEvent(_ input: String) -> PluginResponse {
        guard let obj = json(input), let name = obj["name"] as? String else {
            return PluginResponse(ok: false, output: "expected {name, params?}")
        }
        let params = obj["params"] as? [String: Any]
        Analytics.logEvent(name, parameters: params)
        return PluginResponse(ok: true, output: "")
    }

    private static func setUserProperty(_ input: String) -> PluginResponse {
        guard let obj = json(input), let name = obj["name"] as? String else {
            return PluginResponse(ok: false, output: "expected {name, value}")
        }
        Analytics.setUserProperty(obj["value"] as? String, forName: name)
        return PluginResponse(ok: true, output: "")
    }

    private static func setEnabled(_ input: String) -> PluginResponse {
        let on = (input == "true")
        Analytics.setAnalyticsCollectionEnabled(on)
        Crashlytics.crashlytics().setCrashlyticsCollectionEnabled(on)
        return PluginResponse(ok: true, output: "")
    }

    private static func recordError(_ input: String) -> PluginResponse {
        let obj = json(input)
        let message = (obj?["message"] as? String) ?? "error"
        let domain = (obj?["domain"] as? String) ?? "mobiler.analytics"
        let error = NSError(domain: domain, code: 0, userInfo: [NSLocalizedDescriptionKey: message])
        Crashlytics.crashlytics().record(error: error)
        return PluginResponse(ok: true, output: "")
    }

    private static func json(_ input: String) -> [String: Any]? {
        guard let data = input.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return obj
    }
}

File: mobiler/plugins/analytics/android/AnalyticsPlugin.kt
package {{PACKAGE}}

import {{PACKAGE_SHARED_TYPES}}.PluginResponse

import android.app.Application
import android.os.Bundle
import com.google.firebase.FirebaseApp
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.crashlytics.FirebaseCrashlytics
import org.json.JSONObject

// analytics (free, bundled, EXPERIMENTAL). Product analytics + automatic crash capture via Firebase
// Analytics + Crashlytics. Fire-and-forget request/response (no stream). Crash capture is AUTOMATIC
// once Firebase is initialized (it auto-inits on Android via google-services + the firebase-common
// ContentProvider — no launch hook needed). Ops (input is JSON unless noted):
//   logEvent {name, params?}   setUserId <id>   setUserProperty {name, value}   setEnabled "true"|"false"
//   log <message>              recordError {message, domain?}                    testCrash ""  (dev only)
//
// Needs google-services.json in Android/app/ — the google-services Gradle plugin reads it at build time
// and the build FAILS without it.
class AnalyticsPlugin(private val application: Application) : MobilerPlugin {
    private val analytics by lazy { FirebaseAnalytics.getInstance(application) }
    private val crashlytics by lazy { FirebaseCrashlytics.getInstance() }

    override suspend fun handle(op: String, input: String): PluginResponse {
        if (FirebaseApp.getApps(application).isEmpty()) {
            return PluginResponse(false, "Firebase not configured — add google-services.json")
        }
        return when (op) {
            "logEvent" -> logEvent(input)
            "setUserId" -> {
                analytics.setUserId(input.ifEmpty { null })
                PluginResponse(true, "")
            }
            "setUserProperty" -> setUserProperty(input)
            "setEnabled" -> setEnabled(input == "true")
            "log" -> {
                crashlytics.log(input)
                PluginResponse(true, "")
            }
            "recordError" -> recordError(input)
            // Deliberate crash to verify Crashlytics wiring — the report uploads on the NEXT launch.
            "testCrash" -> throw RuntimeException("analytics testCrash")
            else -> PluginResponse(false, "unknown op '$op'")
        }
    }

    private fun logEvent(input: String): PluginResponse {
        val obj = runCatching { JSONObject(input) }.getOrNull()
            ?: return PluginResponse(false, "invalid input JSON")
        val name = obj.optString("name").ifEmpty { return PluginResponse(false, "expected {name, params?}") }
        val bundle = Bundle()
        obj.optJSONObject("params")?.let { params ->
            for (key in params.keys()) {
                when (val v = params.get(key)) {
                    is String -> bundle.putString(key, v)
                    is Int -> bundle.putLong(key, v.toLong())
                    is Long -> bundle.putLong(key, v)
                    is Double -> bundle.putDouble(key, v)
                    is Boolean -> bundle.putString(key, v.toString())
                    else -> bundle.putString(key, v.toString())
                }
            }
        }
        analytics.logEvent(name, bundle)
        return PluginResponse(true, "")
    }

    private fun setUserProperty(input: String): PluginResponse {
        val obj = runCatching { JSONObject(input) }.getOrNull()
            ?: return PluginResponse(false, "invalid input JSON")
        val name = obj.optString("name").ifEmpty { return PluginResponse(false, "expected {name, value}") }
        analytics.setUserProperty(name, obj.optString("value").ifEmpty { null })
        return PluginResponse(true, "")
    }

    private fun setEnabled(on: Boolean): PluginResponse {
        analytics.setAnalyticsCollectionEnabled(on)
        crashlytics.setCrashlyticsCollectionEnabled(on)
        return PluginResponse(true, "")
    }

    private fun recordError(input: String): PluginResponse {
        val obj = runCatching { JSONObject(input) }.getOrNull()
        val message = obj?.optString("message")?.ifEmpty { null } ?: "error"
        crashlytics.recordException(RuntimeException(message))
        return PluginResponse(true, "")
    }
}
